sqlmesh/core/table_diff.py

```diff
@@ -0,0 +1,195 @@
+from __future__ import annotations
+
+import typing as t
+
+import pandas as pd
+from sqlglot import exp
+from sqlglot.dialects.dialect import DialectType
+
+from sqlmesh.utils.pydantic import PydanticModel
+
+if t.TYPE_CHECKING:
+    from sqlmesh.core._typing import TableName
+    from sqlmesh.core.engine_adapter import EngineAdapter
+
+
+class SchemaDiff(PydanticModel, frozen=True):
+    """An object containing the schema difference between a source and target table."""
+
+    source: str
+    target: str
+    source_schema: t.Dict[str, exp.DataType]
+    target_schema: t.Dict[str, exp.DataType]
+
+    @property
+    def added(self) -> t.List[t.Tuple[str, exp.DataType]]:
+        """Added columns."""
+        return [(c, t) for c, t in self.target_schema.items() if c not in self.source_schema]
+
+    @property
+    def removed(self) -> t.List[t.Tuple[str, exp.DataType]]:
+        """Removed columns."""
+        return [(c, t) for c, t in self.source_schema.items() if c not in self.target_schema]
+
+    @property
+    def modified(self) -> t.Dict[str, t.Tuple[exp.DataType, exp.DataType]]:
+        """Columns with modified types."""
+        modified = {}
+        for column in self.source_schema.keys() & self.target_schema.keys():
+            source_type = self.source_schema[column]
+            target_type = self.target_schema[column]
+
+            if source_type != target_type:
+                modified[column] = (source_type, target_type)
+        return modified
+
+
+class SummaryDiff(PydanticModel, frozen=True):
+    """Summary statistics and a sample dataframe."""
+
+    source: str
+    target: str
+    stats: t.Dict[str, float]
+    sample: pd.DataFrame
+
+    @property
+    def s_count(self) -> int:
+        """Count of the source."""
+        return int(self.stats["s_count"])
+
+    @property
+    def t_count(self) -> int:
```

Comment:
Do we need extraction methods for `*_matches` interface too? I don't think that accessing the `stats` dict directly is a good interface.

Reply:
don't have a use case for it quite yet but we can add those